kyaml/fn/framework/matchers.go

```diff
@@ -0,0 +1,342 @@
+		MetaMatcher: func(labels map[string]string, meta yaml.ResourceMeta) bool {
+			if meta.Labels == nil {
+				return false
+			}
+			for k := range labels {
+				if meta.Labels[k] != labels[k] {
+					return false
+				}
+			}
+			return true
+		},
+	}
+}
+
+// AnnotationMatcher matches resources that are annotated with all of the provided key-value pairs.
+// e.g. `AnnotationMatcher(map[string]string{"app": "foo", "env": "prod"})` matches resources
+// annotated app=foo AND env=prod.
+//
+// AnnotationMatcher supports templating.
+// e.g. `AnnotationMatcher(map[string]string{"app": "{{ .AppName}}"})` will match label app=foo if
+// TemplateData is `struct{ AppName string }{ AppName: "foo" }`
+func AnnotationMatcher(ann map[string]string) ResourceTemplateMatcher {
+	return &TemplatedMetaMapMatcher{
+		Templates: ann,
+		MetaMatcher: func(ann map[string]string, meta yaml.ResourceMeta) bool {
+			if meta.Annotations == nil {
+				return false
+			}
+			for k := range ann {
+				if meta.Annotations[k] != ann[k] {
+					return false
+				}
+			}
+			return true
+		},
+	}
+}
+
+// TemplatedMetaMapMatcher is a utility type for constructing matchers that compare resource
+// metadata to a map of (possibly templated) key-value pairs.
+type TemplatedMetaMapMatcher struct {
+	Templates    map[string]string
+	values       map[string]string
+	TemplateData interface{}
+	MetaMatcher  func(names map[string]string, meta yaml.ResourceMeta) bool
+}
+
+// Match parses the resource node's metadata and delegates matching logic to the provided
+// MetaMatcher func. This allows ResourceMatchers build with TemplatedMetaMapMatcher to match
+// against any field in resource metadata.
+func (m *TemplatedMetaMapMatcher) Match(node *yaml.RNode) bool {
+	var err error
+	meta, err := node.GetMeta()
+	if err != nil {
+		return false
+	}
+
+	return m.MetaMatcher(m.values, meta)
+}
+
+// DefaultTemplateData sets TemplateData to the provided default values if it has not already
+// been set.
+func (m *TemplatedMetaMapMatcher) DefaultTemplateData(data interface{}) {
+	if m.TemplateData == nil {
+		m.TemplateData = data
+	}
+}
+
+// InitTemplates is used to render any templates the selector's key-value pairs may contain
+// before the selector is applied. It should be called exactly once per filter
+// operation, before beginning match comparisons.
+func (m *TemplatedMetaMapMatcher) InitTemplates() error {
+	var err error
+	m.values, err = templatizeMap(m.Templates, m.TemplateData)
+	return errors.Wrap(err)
+}
+
+// Filter applies the matcher to a list of items, returning only those that match.
+func (m *TemplatedMetaMapMatcher) Filter(items []*yaml.RNode) ([]*yaml.RNode, error) {
+	// AndSelector or OrSelector doesn't really matter here since there is only one matcher (m).
+	return AndSelector{Matchers: []ResourceMatcher{m}, TemplateData: m.TemplateData}.Filter(items)
+}
+
+func templatizeSlice(values []string, data interface{}) ([]string, error) {
+	if data == nil {
```

Comment:
While true in general, I'm not sure it should be supported in this context. The templates feature exists to support injecting very simple (single-word meta) values from the config API dynamically. If the template weren't using the values, I don't see why it would be a template at all.

Reply:
Fair